### ISSUE TITLE
Develop LeRobotDataset tools

### ISSUE THREAD

**Issue by michel-aractingi** [huggingface/lerobot#2326]
# 🧰 Call for Contributions: Expanding Dataset Tools in `LeRobotDataset`

Currently, the dataset tools provided in [`src/lerobot/datasets/dataset_tools.py`](https://github.com/huggingface/lerobot/blob/main/src/lerobot/datasets/dataset_tools.py) allow users to:

- Delete episodes from a dataset  
- Split a dataset into multiple sub-datasets  
- Merge multiple datasets into one  
- Modify dataset features by removing existing features and/or adding new ones  

You can try these tools by running `lerobot-edit-dataset`  with the appropriate configuration.

While these tools fill an important gap by enabling manipulation of existing datasets, there’s still a lot of functionality missing.  
This issue is a **call to the community** to help develop and expand the dataset tools for `LeRobotDataset`.

---

### 🗒️ 1. Modify language instructions at the episode level
Once a task description is written in the dataset, it’s currently not possible to edit or extend it (e.g., to fix typos or add a more detailed description).  
We could add tools to update or rewrite the natural language instructions associated with specific episodes.

---

### ➕ 2. Add support for `add_feature` in `lerobot_edit_dataset.py`
Check the `EditDatasetConfig` class in the editing script. The `add_feature` option is not supported yet.  
This would make it possible to easily add new computed or external features without manually modifying parquet files.

---

### 🔄 3. Merge datasets with different feature keys
Currently, the merge tool requires all datasets to share the exact same feature keys.  
This limits merging datasets that are semantically similar but differ in naming conventions.

Possible improvements:
- Allow merging datasets with *semantically matching* features using different keys  
  (e.g., `dataset1: "observation.image.camera1"` vs `dataset2: "observation.image.top"`)  
- Introduce a **feature mapping dictionary** to align features across datasets  
- Implement a **`rename_features`** tool to rename features within a dataset. This would be useful both for merging and as another editing tool  
- *(Advanced)* Support merging datasets with different feature sets by creating a union of all features and padding missing ones appropriately.  
  This would be more complex and could be considered a lower priority.

---

## 💡 Open for Ideas

If you’d like to propose a new tool or enhancement, please **comment below** or **open a PR**. Contributions are highly encouraged!

Your help will make `LeRobotDataset` a more flexible and powerful framework for robotic data management 🚀

Original PR: https://github.com/huggingface/lerobot/pull/2100

**Comment by Grigorij-Dudnik** [comment]
The bug I found #2328 is probably related.

**Comment by jackvial** [comment]
@michel-aractingi I'm interested in taking on `1. Modify language instructions at the episode level`
For the primary syntax I'm thinking something like this where the key of the dict is the episode and the value is the new task string

```bash
python -m lerobot.scripts.lerobot_edit_dataset \
    --repo_id lerobot/pusht \
    --operation.type update_tasks \
    --operation.task_mapping '{0: "Pick up the red cube", 1: "Place the red cube"}'
```

and an additional  "find and replace" syntax might also be useful
```bash
python -m lerobot.scripts.lerobot_edit_dataset \
    --repo_id lerobot/pusht \
    --operation.type update_tasks \
    --operation.task_mapping '{"Pick up the red cube": "Place the red cube"}'
```

a less verbose syntax for updating multiple episodes might look like


```bash
python -m lerobot.scripts.lerobot_edit_dataset \
    --repo_id lerobot/pusht \
    --operation.type update_tasks \
    --operation.task_mapping '{[0, 3, 5]: "Pick up the red cube"}'
```

**Comment by jackvial** [comment]
I think a visual editing tool is a pretty important dataset tool for reviewing and cleaning data, I've built [lerobot data studio](https://github.com/jackvial/lerobot-data-studio) and recently updated it to be built on the new dataset tools API which simplified it quite a bit. If there is any interesting from the community to develop that app further I'd be happy to make a PR to move that code to LeRobot which would probably make it easier for more people to get involved. Also happy to discuss and offer feedback on what a new dataset app GUI app might look like. 

The features I think are most useful from lerobot data studio are:
- Select all/unselect all, and select individual episodes
- Adjusting playback speed
- Keyboard shortcuts. Arrow keys for forward and backward through episodes, space bar to start/stop, ctrl+k or similar to toggle episode selection
- Creating the new dataset from selection

The CLI tool works better for most other dataset tooling and I think a GUI should be pretty minimal and not duplicate functionality that is better suited and already available in the CLI tool.

**Comment by helper2424** [comment]
## Additional Feature Requests / Design Considerations

I’d like to highlight a few areas that I believe are increasingly important for LeRobot, especially given the rapid scaling trends in robotics datasets and model training:

- [ ] **Scalability for very large datasets**
- [ ] **Mixed-frequency / asynchronous data (different FPS per modality)**
- [ ] **Full post-collection editability / dataset surgery tools**
- [ ] **Merging multiple datasets (logical + physical)**

---

### ✅ 1. Scalability for Very Large Datasets

Recent announcements — for example, Generalist AI reporting **270k+ hours** of data — suggest that the field is moving toward **hundreds of thousands and soon millions of hours** of robot data.

Link:  
🔗 https://generalistai.com/blog/nov-04-2025-GEN-0

LeRobot is already the best open-source ecosystem for robotics datasets, but it’s important to ask:

**Is the current dataset format ready for such heavy scale?**  
And if not, how can we evolve it to get there?

It would be extremely valuable to ensure LeRobot can:

- handle multi-hundred-TB or petabyte-scale datasets  
- stream efficiently  
- support distributed reading/writing  
- keep metadata/indexes lightweight and scalable  
- avoid bottlenecks when mixing or globally shuffling huge datasets

This would position LeRobot as *the* foundation for future generalist robot models.

---

### ✅ 2. Mixed-Frequency / Asynchronous Data (Different FPS)

Real-world robot data is inherently asynchronous:

- cameras → each stream may have different FPS  
- state/action logs → often irregular timestamps  
- audio → different sampling rate  
- force/torque → another frequency  

Examples and references:  
- Positronic dataset format:  
  🔗 https://github.com/Positronic-Robotics/positronic/blob/main/positronic/dataset/README.md  
- Generalist AI “Harmonic Reasoning” discussion:  
  🔗 https://generalistai.com/blog/nov-04-2025-GEN-0

It would be great if LeRobot supported:

- storing each modality at its **native sampling rate**  
- precise timestamp alignment  
- utilities for **resampling, interpolating, fusing, and aligning** modalities  
- making asynchronous data a **first-class feature** in the dataset spec  

---

### ✅ 3. Full Post-Collection Editability ("Dataset Surgery")

After collection, **no dataset is perfect**. Everyone ends up performing modifications like:

- removing bad frames or episodes  
- fixing numerical or text values  
- replacing broken video segments  
- editing metadata  
- filtering low-quality segments  
- merging or splitting episodes  

Right now this often requires ad-hoc scripting or re-exporting.  
It would be extremely helpful to have **official utilities** for:

- deleting or replacing frames / episodes  
- modifying fields in-place  
- replacing video/audio streams without full rebuild  
- batch transforms  
- fast, consistent index/manifest rebuilding  

This would save huge amounts of development time and reduce errors.

---

### ✅ 4. Dataset Merging (Logical & Physical)

Large-scale multitask training typically uses multiple datasets collected at different times, with different robots, tasks, or data formats.

LeRobot would benefit from supporting two types of merges:

#### **Logical merge**
A unified Dataset object that:

- samples cleanly from several datasets  
- supports weighting / mixing  
- aligns modalities  
- presents a consistent interface to the trainer  

#### **Physical merge**
Tools to produce a new dataset that:

- unifies metadata and versioning  
- harmonizes modality names  
- reindexes episodes  
- creates a coherent, clean dataset on disk

**Comment by jackvial** [comment]
@helper2424 thanks for putting this together, I think that is a good set of goals to strive for when working with large datasets. I've been trying to get setup working with the droid dataset to help me identify were the bottlenecks are with current tools. Downloading via LeRobot datasets failed, git clone failed too, both getting stuck at about 20-30%, I don't think I was hitting my API limits but I did upgrade to HF pro for higher rate limits but that still didn't fix the problem so I think my ISP was throttling my downloads.

What worked in the end was using this script

```python
import os
from huggingface_hub import snapshot_download, logging

# Enable the high-speed Rust backend
os.environ["HF_HUB_ENABLE_HF_TRANSFER"] = "1"

# Enable detailed logging (shows transfer backend + request info)
logging.set_verbosity_debug()

# Download the full dataset repository
local_path = snapshot_download(
    repo_id="lerobot/droid_1.0.1",   # dataset you want to pull
    repo_type="dataset",             # make sure it's a dataset repo
    local_dir="./droid_1.0.1",       # where to store it locally
    resume_download=True,            # continue if interrupted
    max_workers=4,                   # adjust concurrency if your ISP throttles
)

print(f"Dataset downloaded to: {local_path}")
```

From what I gather HF transfer does some special handling that prevents ISP from throttling download speeds? Would be good to hear what others experience has been download these large datasets.

Next I am going to try loading this in my LeRobot data studio app and can also try do some operations with the new LeRobot dataset tools

**Comment by masato-ka** [comment]
I'm suggesting easy but important feature that show the basic informaion dataset.
This function provide function that show the number of episode, feature_key and dimention, all task instruction.
We need to well know fundamental information of the datasets before editting it.

**Comment by hairuoliu1** [comment]
> ## 其他功能需求/设计考虑
> 我想重点强调几个我认为对乐视机器人而言日益重要的领域，尤其是在机器人数据集和模型训练快速扩展的趋势下：
> 
> * [ ]  **针对超大型数据集的可扩展性**[ ]  **混合频/异步数据（不同模态的帧速率不同）**[ ]  **完整的后期编辑/数据集操作工具**[ ]  **合并多个数据集（逻辑数据集+物理数据集）**
> 
> ### ✅ 1. 可扩展性，适用于超大型数据集
> 最近的公告——例如，通用人工智能报告称拥有**超过 27 万小时**的数据——表明该领域正在朝着**数十万小时，很快将达到数百万小时**的机器人数据迈进。
> 
> 链接： 🔗 https://generalistai.com/blog/nov-04-2025-GEN-0
> 
> LeRobot 已经是目前最好的机器人数据集开源生态系统，但还有一个问题值得探讨：
> 
> **当前的数据集格式是否能够应对如此庞大的规模？** 如果不能，我们该如何改进才能使其达到要求？
> 
> 确保乐视机器人能够做到以下几点将极其重要：
> 
> * 处理数百TB或PB级数据集
> * 高效流
> * 支持分布式阅读/写作
> * 保持元数据/索引轻量级且可扩展
> * 避免在混合或全局重组大型数据集时出现瓶颈
> 
> 这将使 LeRobot 成为未来通用机器人模型的基础_。_
> 
> ### ✅ 2. 混合频率/异步数据（不同帧率）
> 现实世界中的机器人数据本质上是异步的：
> 
> * 摄像头 → 每个视频流的帧率可能不同
> * 状态/操作日志 → 通常带有不规则的时间戳
> * 音频 → 不同的采样率
> * 力/扭矩 → 另一个频率
> 
> 示例和参考文献：
> 
> * Positronic 数据集格式：
>   🔗 https://github.com/Positronic-Robotics/positronic/blob/main/positronic/dataset/README.md
> * 通用人工智能“调和推理”讨论：
>   🔗 https://generalistai.com/blog/nov-04-2025-GEN-0
> 
> 如果 LeRobot 支持以下功能就太好了：
> 
> * 以每种模态的**原始采样率存储**
> * 精确时间戳对齐
> * 用于**重采样、插值、融合和对齐**模态的实用程序
> * 将异步数据作为数据集规范中的**一流特性**
> 
> ### ✅ 3. 完全的后处理可编辑性（“数据集操作”）
> 收集数据之后，**没有哪个数据集是完美的**。每个人最终都会进行一些修改，例如：
> 
> * 删除错误帧或片段
> * 修正数值或文本值
> * 替换损坏的视频片段
> * 编辑元数据
> * 过滤低质量片段
> * 合并或拆分剧集
> 
> 目前，这通常需要临时编写脚本或重新导出。如果能有**官方的实用程序** 来实现以下功能，将会非常有帮助：****
> 
> * 删除或替换帧/集
> * 就地修改字段
> * 无需完全重建即可替换视频/音频流
> * 批量转换
> * 快速、一致的索引/清单重建
> 
> 这将节省大量的开发时间并减少错误。
> 
> ### ✅ 4. 数据集合并（逻辑合并和物理合并）
> 大规模多任务训练通常使用在不同时间收集的多个数据集，这些数据集可能包含不同的机器人、任务或数据格式。
> 
> LeRobot 可以通过支持两种类型的合并而受益：
> 
> #### **逻辑合并**
> 一个统一的数据集对象，它：
> 
> * 从多个数据集中干净地提取样本
> * 支持加重/混合
> * 调整模式
> * 为培训师提供一致的界面
> 
> #### **物理合并**
> 用于生成新数据集的工具：
> 
> * 统一元数据和版本控制
> * 统一模态名称
> * 重新索引剧集
> * 在磁盘上创建一个连贯、干净的数据集。

I agree you!

**Comment by jackvial** [comment]
> [@michel-aractingi](https://github.com/michel-aractingi) I'm interested in taking on `1. Modify language instructions at the episode level` For the primary syntax I'm thinking something like this where the key of the dict is the episode and the value is the new task string
> 
> python -m lerobot.scripts.lerobot_edit_dataset \
>     --repo_id lerobot/pusht \
>     --operation.type update_tasks \
>     --operation.task_mapping '{0: "Pick up the red cube", 1: "Place the red cube"}'
> 
> and an additional "find and replace" syntax might also be useful
> 
> python -m lerobot.scripts.lerobot_edit_dataset \
>     --repo_id lerobot/pusht \
>     --operation.type update_tasks \
>     --operation.task_mapping '{"Pick up the red cube": "Place the red cube"}'
> 
> a less verbose syntax for updating multiple episodes might look like
> 
> python -m lerobot.scripts.lerobot_edit_dataset \
>     --repo_id lerobot/pusht \
>     --operation.type update_tasks \
>     --operation.task_mapping '{[0, 3, 5]: "Pick up the red cube"}'

Hey all, I've not be able to find the time to take this further so please anyone feel free to pick this up

**Comment by masato-ka** [comment]
I would like to adding new feature to show info of the datasets (#2917 ).

I believe there is no easy way to check dataset information from the command line. This feature provides that capability. I would like to hear your opinions on whether it should be a subset of lerobot-edit-dataset or an independent feature.

**Comment by masato-ka** [comment]
I propose and open PR that exporting each episode video function.  We would like to need to each episode video to share our evaluation result with community or something. Dataset V3 some episode encode to one video file.
We can be easily separating each episode videofile by this PR(#2983 ).

**Comment by boysack** [comment]
I've written the code to add support for `add_feature` in `lerobot_edit_dataset.py`, and I'm proposing these changes in PR #3260. Specifically, the proposed solution offers optimized RAM usage to prevent memory overflows during heavy data loads. It also introduces a multithreaded approach to add video sources, ensuring both modalities are correctly split into chunks and files.

**Comment by jashshah999** [comment]
We built [lerobot-doctor](https://github.com/jashshah999/lerobot-doctor) as a complementary diagnostics/repair toolkit addressing several needs here:

- `merge-check` — validates dataset merges (catches feature mismatches, index conflicts)
- `trim` — removes idle frames (addresses data curation)
- `score` — per-episode quality scoring for curation
- `fix` — auto-repairs broken indices, timestamps, NaN values
- `gate --policy act/diffusion/smolvla` — pre-training compatibility check

pip-installable, no lerobot dependency. Already integrated into the official lerobot-dataset-visualizer (PR #57).

https://github.com/jashshah999/lerobot-doctor

**Comment by delta003** [comment]
Not sure how relevant the large datasets discussion still is, but I was recently optimizing a few things for one of our customers, which led me on a side quest to explore some dataset format changes and had some good results https://claude.ai/code/artifact/b3870281-def9-4452-ba90-7ad8d779a711. The main thing is that the format has container formats (mp4) within a container format (lerobot v3), and mp4 is not nearly as nice as parquet. The experiment gets rid of mp4 in favour of GOP units in Vortex (https://vortex.dev/) files, but it can go as far as fusing the whole dataset format into just Vortex files. Happy to help if/when you're looking at v4 or https://spiraldb.com/deep-dives/video has some design techniques you can use.

**Comment by easyrider11** [comment]
Picking up item 3 (merging datasets with different feature keys).

The `rename_features` half is done and working against current `main`. It continues @IamPhytan's #2435, which has been a draft since January and no longer rebases (those files moved +1684/-337 since its base), so I ported the design over rather than rebasing.

Beyond renaming the parquet column it also needed the dataset-level stats, the per-episode `stats/<feature>/*` columns, video file relocation (video files live under a directory named after the feature key), and the per-video `info` block. The first two were dropping data silently rather than erroring.

Shape of it:

- `modify_features(..., rename_features={"old": "new"})`, composing with add/remove in the same single-pass copy
- `rename_feature(dataset, mapping, ...)` next to `add_features` / `remove_feature`
- `lerobot-edit-dataset --operation.type rename_feature --operation.feature_mapping '{"old": "new"}'`
- 10 new tests incl. a video-feature rename; `pytest tests/datasets/test_dataset_tools.py` → 77 passed, ruff clean

Checking before I open anything: would you rather have this as a fresh PR (with @IamPhytan as co-author), or should I push onto #2435? I've asked there too.

The second half of item 3 — `merge_datasets` taking a feature mapping directly — becomes straightforward once renaming exists. I'd do that as a follow-up rather than bundle it, unless you'd prefer one PR.

**Comment by easyrider11** [comment]
Answering the call for dataset tools — I built a **contract linter**: https://github.com/easyrider11/lerobot-dataset-lint

```bash
uvx lerobot-dataset-lint lerobot/pusht   # any hub repo id or local dir
```

19 rules: stats ↔ data ↔ metadata consistency (stale dims/envelopes, zero-std), dropped frames, v3 global-index tiling, video-timeline coverage (no decoding needed), NaN/Inf, dead recordings, teleop clipping, and a gripper-convention report (mixed conventions across episodes = error). No torch or lerobot import — it fetches meta plus only the parquet files holding the sampled episodes, so linting a multi-GB dataset costs megabytes. Supports v2.x and v3.0, CI-friendly exit codes, `--json`.

Two findings maintainers might care about:

1. Official v3 datasets disagree on task association: `lerobot/pusht` stores `tasks` in the episodes index, `lerobot/libero` only has per-frame `task_index`. The linter handles both — which one is canonical?
2. A first community sweep found real issues, e.g. an action dim sitting at its bound in 64% of frames (teleop clipping) in a public dataset.

Happy to adapt the rule set or donate the tool in whatever shape fits the roadmap. Apache-2.0.

**Comment by psteinb** [comment]
I agree with respect of making lerobot datasets ready for scaling. 

I am currently looking into this dataset:
https://huggingface.co/datasets/IPEC-COMMUNITY/bridge_orig_lerobot
And it turns out, the setup of the dataset (~270k files with an average of 10kB per file) exposes a performance penalty for our parallel filesystem on the cluster. So tooling that would e.g. allow us to increase the chunk size to something larger than 4MB would be really helpful.
